gui_interactions.md

```diff
@@ -0,0 +1,491 @@
+# GUI interactions
+
+This document is meant to inform and guide the development of various GUI
+interactions for Ignition Gazebo.
+
+## Overview
+
+Ignition Gazebo's graphical interface's high-level goals are to:
+
+1. **Introspect** a running or recorded simulation
+2. **Interact** with a running or recorded simulation
+3. **Edit** simulation worlds and robots
+
+**Introspection** allows users to see how objects are positioned through a 3D
+view, the precise position of objects through text widgets, how velocities
+are changing over time through plots, etc.
+
+**Interaction** allows users to apply forces to a robot while it moves,
+to teleport objects in the scene while simulation is running, etc.
+
+**Editing** allows users to arrange world environments, assemble new
+robots and objects, modify existing objects, etc.
+
+The main difference between interaction and editing is that the former must
+happen while the simulation is running, while the latter doesn't. But in several
+aspects, their functionalities overlap. For example, teleporting an object may
+be useful while the simulation is running to test whether a robot will keep
+following it. And teleporting objects is also essential to creating a new world
+environment.
+
+It could be argued that every aspect of editing could be helpful for
+interaction. Even actions like changing a link's mass or adding a new link while
+physics is running will be helpful to some user. However, it's important to keep
+in mind that handling interactions is considerably more complicated than
+handling editing, primarily because physics engines are usually designed to deal
+with bodies that have immutable kinematics and dynamics.
+
+With that in mind, this design document proposes the separation of features into
+interaction and editing. The central difference between them being that editing
+features can't happen while physics is running, while interaction features can
+happen any time. In other words, we will have two separate modes of operation:
+
+* **Edit mode**: whenever simulation is **paused**
+* **Interactive mode**: whenever simulation is **playing**
+
+Different modes of operation are common among software which allows both
+creating and running content. Think about how slides can't be edited while
+they're being presented, or how video editors will show short previews while
+editing a video, but once the video is generated it can no longer be edited.
+Closer to the robotics world, we have game engines which often have a clearly
+separate run step for the game to be tested.
+
+While implementing these different modes of operation, it's important to:
+
+* Keep the user interface consistent across modes.
```

Comment:
I like consistency, but I think it should also be easy to tell which mode you are in, without having to notice the state of the small play / pause icon or watch for changing text values in different widgets. Having completely different user interfaces would make it easy to see which mode you are in, though the lack of consistency could be disorienting or confusing. I'm sure there's a compromise to be made here that would allow a bold / obvious change in the U/I to quickly show which mode you are in.

Reply:
That's a very good point about making the current mode clearer. I'll add something about this.

When I talk about consistency, I was mainly thinking of how different the two interfaces are on Gazebo-classic, see a link's properties for example. I think it would be better for the user if they only had to get familiar with a single style. 

Simulation | Editor
-- | --
![classic_pose](https://user-images.githubusercontent.com/5751272/82262581-0aef3300-9916-11ea-96d5-bd27a647c1b9.png) | ![editor_pose](https://user-images.githubusercontent.com/5751272/82262600-104c7d80-9916-11ea-811d-dae620586fb9.png)

Reply:
3edf568

Reply:
👍